In [26]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler, StandardScaler
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
data = pd.read_csv("spanish-cities-energy-consumption/weather_features.csv")
data.head()

dt_iso city_name     temp  temp_min  temp_max  pressure  \
0  2015-01-01 00:00:00+01:00  Valencia  270.475   270.475   270.475      1001   
1  2015-01-01 01:00:00+01:00  Valencia  270.475   270.475   270.475      1001   
2  2015-01-01 02:00:00+01:00  Valencia  269.686   269.686   269.686      1002   
3  2015-01-01 03:00:00+01:00  Valencia  269.686   269.686   269.686      1002   
4  2015-01-01 04:00:00+01:00  Valencia  269.686   269.686   269.686      1002   

   humidity  wind_speed  wind_deg  rain_1h  rain_3h  snow_3h  clouds_all  \
0        77           1        62      0.0      0.0      0.0           0   
1        77           1        62      0.0      0.0      0.0           0   
2        78           0        23      0.0      0.0      0.0           0   
3        78           0        23      0.0      0.0      0.0           0   
4        78           0        23      0.0      0.0      0.0           0   

   weather_id weather_main weather_description weather_icon  
0         800        clear        sky is clear          01n  
1         800        clear        sky is clear          01n  
2         800        clear        sky is clear          01n  
3         800        clear        sky is clear          01n  
4         800        clear        sky is clear          01n

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dt_iso</th>
      <th>city_name</th>
      <th>temp</th>
      <th>temp_min</th>
      <th>temp_max</th>
      <th>pressure</th>
      <th>humidity</th>
      <th>wind_speed</th>
      <th>wind_deg</th>
      <th>rain_1h</th>
      <th>rain_3h</th>
      <th>snow_3h</th>
      <th>clouds_all</th>
      <th>weather_id</th>
      <th>weather_main</th>
      <th>weather_description</th>
      <th>weather_icon</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2015-01-01 00:00:00+01:00</td>
      <td>Valencia</td>
      <td>270.475</td>
      <td>270.475</td>
      <td>270.475</td>
      <td>1001</td>
      <td>77</td>
      <td>1</td>
      <td>62</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0</td>
      <td>800</td>
      <td>clear</td>
      <td>sky is clear</td>
      <td>01n</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2015-01-01 01:00:00+01:00</td>
      <td>Valencia</td>
      <td>270.475</td>
      <td>270.475</td>
      <td>270.475</td>
      <td>1001</td>
      <td>77</td>
      <td>1</td>
      <td>62</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0</td>
      <td>800</td>
      <td>clear</td>
      <td>sky is clear</td>
      <td>01n</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2015-01-01 02:00:00+01:00</td>
      <td>Valencia</td>
      <td>269.686</td>
      <td>269.686</td>
      <td>269.686</td>
      <td>1002</td>
      <td>78</td>
      <td>0</td>
      <td>23</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0</td>
      <td>800</td>
      <td>clear</td>
      <td>sky is clear</td>
      <td>01n</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2015-01-01 03:00:00+01:00</td>
      <td>Valencia</td>
      <td>269.686</td>
      <td>269.686</td>
      <td>269.686</td>
      <td>1002</td>
      <td>78</td>
      <td>0</td>
      <td>23</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0</td>
      <td>800</td>
      <td>clear</td>
      <td>sky is clear</td>
      <td>01n</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2015-01-01 04:00:00+01:00</td>
      <td>Valencia</td>
      <td>269.686</td>
      <td>269.686</td>
      <td>269.686</td>
      <td>1002</td>
      <td>78</td>
      <td>0</td>
      <td>23</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0</td>
      <td>800</td>
      <td>clear</td>
      <td>sky is clear</td>
      <td>01n</td>
    </tr>
  </tbody>
</table>
</div>

In [27]:
data.describe()

temp       temp_min       temp_max      pressure  \
count  178396.000000  178396.000000  178396.000000  1.783960e+05   
mean      289.618605     288.330442     291.091267  1.069261e+03   
std         8.026199       7.955491       8.612454  5.969632e+03   
min       262.240000     262.240000     262.240000  0.000000e+00   
25%       283.670000     282.483602     284.650000  1.013000e+03   
50%       289.150000     288.150000     290.150000  1.018000e+03   
75%       295.150000     293.730125     297.150000  1.022000e+03   
max       315.600000     315.150000     321.150000  1.008371e+06   

            humidity    wind_speed       wind_deg        rain_1h  \
count  178396.000000  178396.00000  178396.000000  178396.000000   
mean       68.423457       2.47056     166.591190       0.075492   
std        21.902888       2.09591     116.611927       0.398847   
min         0.000000       0.00000       0.000000       0.000000   
25%        53.000000       1.00000      55.000000       0.000000   
50%        72.000000       2.00000     177.000000       0.000000   
75%        87.000000       4.00000     270.000000       0.000000   
max       100.000000     133.00000     360.000000      12.000000   

             rain_3h        snow_3h     clouds_all     weather_id  
count  178396.000000  178396.000000  178396.000000  178396.000000  
mean        0.000380       0.004763      25.073292     759.831902  
std         0.007288       0.222604      30.774129     108.733223  
min         0.000000       0.000000       0.000000     200.000000  
25%         0.000000       0.000000       0.000000     800.000000  
50%         0.000000       0.000000      20.000000     800.000000  
75%         0.000000       0.000000      40.000000     801.000000  
max         2.315000      21.500000     100.000000     804.000000

In [28]:
for col, dtype in data.dtypes.items():
    print(f"Column '{col}' has data type: {dtype}")

Column 'dt_iso' has data type: object
Column 'city_name' has data type: object
Column 'temp' has data type: float64
Column 'temp_min' has data type: float64
Column 'temp_max' has data type: float64
Column 'pressure' has data type: int64
Column 'humidity' has data type: int64
Column 'wind_speed' has data type: int64
Column 'wind_deg' has data type: int64
Column 'rain_1h' has data type: float64
Column 'rain_3h' has data type: float64
Column 'snow_3h' has data type: float64
Column 'clouds_all' has data type: int64
Column 'weather_id' has data type: int64
Column 'weather_main' has data type: object
Column 'weather_description' has data type: object
Column 'weather_icon' has data type: object


In [29]:
data.isna().sum()

dt_iso                 0
city_name              0
temp                   0
temp_min               0
temp_max               0
pressure               0
humidity               0
wind_speed             0
wind_deg               0
rain_1h                0
rain_3h                0
snow_3h                0
clouds_all             0
weather_id             0
weather_main           0
weather_description    0
weather_icon           0
dtype: int64

In [30]:
print(data.isnull().values.any())

False


In [31]:
# for col in z_scores.columns:
#     # Plot Histogram of Z-scores
#     plt.figure(figsize=(12, 5))

#     # Histogram plot
#     plt.subplot(1, 2, 1)
#     plt.hist(z_scores[col], bins=100, edgecolor='k', alpha=0.7)
#     plt.axvline(threshold, color='r', linestyle='dashed', linewidth=1.5, label=f'Outlier Threshold (+{threshold})')
#     plt.axvline(-threshold, color='r', linestyle='dashed', linewidth=1.5, label=f'Outlier Threshold (-{threshold})')
#     plt.title(f'Histogram of Z-scores {col}')
#     plt.xlabel('Z-score')
#     plt.ylabel('Frequency')
#     plt.yscale('log')
#     plt.legend()

#     # Scatter Plot with Z-scores
#     plt.subplot(1, 2, 2)
#     plt.scatter(range(len(z_scores[col])), z_scores[col], c=['red' if outlier else 'blue' for outlier in outliers[col]], alpha=0.7)
#     plt.axhline(threshold, color='r', linestyle='dashed', linewidth=1.5)
#     plt.axhline(-threshold, color='r', linestyle='dashed', linewidth=1.5)
#     plt.title(f'Scatter Plot of Z-scores {col}')
#     plt.xlabel('Index')
#     plt.ylabel('Z-score')
#     plt.grid(True)

#     # Highlight outliers
#     for i, (score, is_outlier) in enumerate(zip(z_scores[col], outliers)):
#         if is_outlier:
#             plt.text(i, score, f'{score:.2f}', fontsize=9, color='red', ha='left', va='bottom')

#     plt.tight_layout()
#     plt.show()

# Feature Scaling

Min-Max-Scaling

In [32]:
numerical_cols = data.select_dtypes(include=['float64', 'int64']).columns


min_max_scaler = MinMaxScaler()


data_minmax_scaled = data.copy()
data_minmax_scaled[numerical_cols] = min_max_scaler.fit_transform(data[numerical_cols])


data_minmax_scaled.head()


# data_minmax_scaled[numerical_cols].hist(figsize=(10, 10))
# plt.suptitle('Min-Max Scaled Data')
# plt.show()




dt_iso city_name      temp  temp_min  temp_max  \
0  2015-01-01 00:00:00+01:00  Valencia  0.154329  0.155642  0.139790   
1  2015-01-01 01:00:00+01:00  Valencia  0.154329  0.155642  0.139790   
2  2015-01-01 02:00:00+01:00  Valencia  0.139543  0.140730  0.126396   
3  2015-01-01 03:00:00+01:00  Valencia  0.139543  0.140730  0.126396   
4  2015-01-01 04:00:00+01:00  Valencia  0.139543  0.140730  0.126396   

   pressure  humidity  wind_speed  wind_deg  rain_1h  rain_3h  snow_3h  \
0  0.000993      0.77    0.007519  0.172222      0.0      0.0      0.0   
1  0.000993      0.77    0.007519  0.172222      0.0      0.0      0.0   
2  0.000994      0.78    0.000000  0.063889      0.0      0.0      0.0   
3  0.000994      0.78    0.000000  0.063889      0.0      0.0      0.0   
4  0.000994      0.78    0.000000  0.063889      0.0      0.0      0.0   

   clouds_all  weather_id weather_main weather_description weather_icon  
0         0.0    0.993377        clear        sky is clear          01n  
1         0.0    0.993377        clear        sky is clear          01n  
2         0.0    0.993377        clear        sky is clear          01n  
3         0.0    0.993377        clear        sky is clear          01n  
4         0.0    0.993377        clear        sky is clear          01n

Z-score Normalization (Standardization):

In [33]:

standard_scaler = StandardScaler()
data_standard_scaled = data.copy()
data_standard_scaled[numerical_cols] = standard_scaler.fit_transform(data[numerical_cols])

data_standard_scaled.head()


# data_standard_scaled[numerical_cols].hist(figsize=(10, 10))
# plt.suptitle('Standardized Data')
# plt.show()


\begin{tabular}{lllrrrrrrrrrrrrlll}
\toprule
 & dt_iso & city_name & temp & temp_min & temp_max & pressure & humidity & wind_speed & wind_deg & rain_1h & rain_3h & snow_3h & clouds_all & weather_id & weather_main & weather_description & weather_icon \\
\midrule
0 & 2015-01-01 00:00:00+01:00 & Valencia & -2.385146 & -2.244424 & -2.393780 & -0.011435 & 0.391572 & -0.701635 & -0.896919 & -0.189276 & -0.052105 & -0.021396 & -0.814755 & 0.369420 & clear & sky is clear & 01n \\
1 & 2015-01-01 01:00:00+01:00 & Valencia & -2.385146 & -2.244424 & -2.393780 & -0.011435 & 0.391572 & -0.701635 & -0.896919 & -0.189276 & -0.052105 & -0.021396 & -0.814755 & 0.369420 & clear & sky is clear & 01n \\
2 & 2015-01-01 02:00:00+01:00 & Valencia & -2.483450 & -2.343601 & -2.485392 & -0.011267 & 0.437229 & -1.178756 & -1.231363 & -0.189276 & -0.052105 & -0.021396 & -0.814755 & 0.369420 & clear & sky is clear & 01n \\
3 & 2015-01-01 03:00:00+01:00 & Valencia & -2.483450 & -2.343601 & -2.485392 & -0.011267 & 0.